compiler/dcalc/ast.ml

```diff
@@ -103,28 +105,88 @@ type unop =
 
 type operator = Ternop of ternop | Binop of binop | Unop of unop
 
-type expr =
-  | EVar of expr Bindlib.var Pos.marked
-  | ETuple of expr Pos.marked list * struct_name option
+(** Some structures used for type inference *)
+module Infer = struct
+  module Any =
+    Utils.Uid.Make
+      (struct
+        type info = unit
+
+        let format_info fmt () = Format.fprintf fmt "any"
+      end)
+      ()
+
+  type unionfind_typ = typ Marked.pos UnionFind.elem
+  (** We do not reuse {!type: Dcalc.Ast.typ} because we have to include a new
+      [TAny] variant. Indeed, error terms can have any type and this has to be
+      captured by the type sytem. *)
+
+  and typ =
+    | TLit of typ_lit
+    | TArrow of unionfind_typ * unionfind_typ
+    | TTuple of unionfind_typ list * StructName.t option
+    | TEnum of unionfind_typ list * EnumName.t
+    | TArray of unionfind_typ
+    | TAny of Any.t
+
+  let rec typ_to_ast (ty : unionfind_typ) : marked_typ =
+    let ty, pos = UnionFind.get (UnionFind.find ty) in
+    match ty with
+    | TLit l -> TLit l, pos
+    | TTuple (ts, s) -> TTuple (List.map typ_to_ast ts, s), pos
+    | TEnum (ts, e) -> TEnum (List.map typ_to_ast ts, e), pos
+    | TArrow (t1, t2) -> TArrow (typ_to_ast t1, typ_to_ast t2), pos
+    | TAny _ -> TAny, pos
+    | TArray t1 -> TArray (typ_to_ast t1), pos
+
+  let rec ast_to_typ (ty : marked_typ) : unionfind_typ =
+    let ty' =
+      match Marked.unmark ty with
+      | TLit l -> TLit l
+      | TArrow (t1, t2) -> TArrow (ast_to_typ t1, ast_to_typ t2)
+      | TTuple (ts, s) -> TTuple (List.map (fun t -> ast_to_typ t) ts, s)
+      | TEnum (ts, e) -> TEnum (List.map (fun t -> ast_to_typ t) ts, e)
+      | TArray t -> TArray (ast_to_typ t)
+      | TAny -> TAny (Any.fresh ())
+    in
+    UnionFind.make (Marked.same_mark_as ty' ty)
+end
+
+type untyped = { pos : Pos.t } [@@ocaml.unboxed]
+type typed = { pos : Pos.t; ty : Infer.unionfind_typ }
```

Comment:
Can't you change the `Untyped | Typed` marking to a `Untyped | TypingInProgress | Typed` enumeration ?

Reply:
Yes! That's the idea. Maybe we can even manage to keep `TypingInProgress` hidden within the `Typing` module. It will be cleaner on the later passes, but adds an additional mapping on the whole AST.